### PR TITLE
Add demonstration generated code for ordered maps in ygot

### PR DESCRIPTION
**Pull request by wenovus**
Once this is approved development for code generation can begin.

### CODE REVIEW

File: gogen/internal/gotypes/ordered_map.go
Comment (DanG100) [inline gogen/internal/gotypes/ordered_map.go:121]: i don't think Update and AppendNew would  be that useful, what use cases did you have in mind for them.

I'd rather have GetOrAppend(key) *RoutingPolicyDefinition

Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go:121]: I added `Update` as a basic dictionary operation. I don't have a specific use case in mind. I think it's ok to remove it if we want to keep to specific use cases.

`AppendNew` was added because Scott mentioned the current `New` will be useful alongside `Append`. This would be the equivalent for ordered maps.

As for `GetOrAppend`, this was in the draft I proposed originally since it follows existing conventions with unordered maps. I received feedback that this might be confusing because they want to know whether it was a Get or Append, since with ordered maps each element in general (e.g. BGP policies) cares about what other elements already exist, whereas in unordered maps each element often does not care about the existence of other elements (e.g. interfaces). So splitting it into `Get` and `AppendNew` helps the user write code that's less ambiguous. The stakeholders mentioned `New` and `Append` as the primary functionalities that they would use.

Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go:121]: I'll try to get a code review from the users.

**Review by bormanp — CHANGES_REQUESTED** [review]
Sorry, I just realized my comments were still pending :-(

File: gogen/internal/gotypes/ordered_map.go
Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go:64]: I can add it. Although the reason why I don't think it's important to add this now is because ygot-generated code is not currently thread-safe. There are instances of slice types, and slices by themselves are not thread-safe -- they would have to be encapsulated just like this type for them to be so. The other pointer types are technically ok because they're benign data races, and perhaps in Go these are truly benign, although I'm not 100% sure.

File: gogen/internal/gotypes/ordered_map.go
Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go:77]: Sigh, I think similar to the above, ygot was not designed to be thread-safe, and so for the current map and slice types users would be calling `len()` just like this function does. IMO it doesn't seem very useful to the user if one very-rare type in the GoStruct is thread-safe while everything else needs a lock anyways. That's why I'm not sure about the value of making this thread-safe at the current time.

File: gogen/internal/gotypes/ordered_map.go
Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go:82]: Sorry I copied this comment from somewhere else and didn't update it, let me do that.

File: gogen/internal/gotypes/ordered_map.go
Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go:91]: Actually Go allows you to index into a nil map and it returns the zero value so no need to call `init` here.

I will change it to short-circuit when the key doesn't exist. Due to the variables being private `keys` and `valueMap` must always be consistent.

File: gogen/internal/gotypes/ordered_map.go
Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go:47]: Actually I will hold off pending our discussion below, I'm still not convinced it's the right time to add this since the rest of ygot doesn't deal with thread-safety. If we do want thread-safety I think we would want another suite of tests. I also think this can be added in a separate PR.

File: gogen/internal/gotypes/ordered_map.go
Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go]: So I got feedback that `Append` is probably all they will need. For `AppendNew` I think given this API exists in the existing API method set, that it does not subtract value to have. I've removed `Update` though since I agree it might be a stretch for people to use it. PTAL

File: gogen/internal/gotypes/ordered_map.go
Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go]: Done.

File: gogen/internal/gotypes/ordered_map.go
Comment (DanG100) [reply gogen/internal/gotypes/ordered_map.go]: here's my thoughts. 
1. GetOrCreate is by far the most commonly used function (at least in featureprofiles). Because it doesn't return an error, it is really convenient to use.  
2. Append: is very rarely used, though it's a still valuable and should exist
3. AppendNew: I don't see this in the existing API set, and personally don't see when i'd use it.

i'm fine with the API as is.

**Review by DanG100 — APPROVED** [review]

File: gogen/internal/gotypes/ordered_map.go
Comment (wenovus) [reply gogen/internal/gotypes/ordered_map.go]: I agree with 1&2 from a current-API perspective, the current PR reflects iteration after feedback. Ordered-maps are different than unordered-maps and IMO this method set has greater usability than simply copying what the existing unordered-map API does.

For 3 AppendNew corresponds to New for unordered maps.

**Pull request merged**
